src/inline-assembly.md

```diff
@@ -684,6 +680,14 @@ and the rules for assembly may include thousands of pages of architectural refer
 Programmers should exercise appropriate care, as invoking this `unsafe` capability comes with
 assuming the responsibility of not violating rules of both the compiler or the architecture.
 
+r[asm.validity.stateful]
+If inline assembly includes any "stateful" directive that modifies how subsequent assembly is processed, the block must undo the effects of any such directives before the inline assembly ends. If this constraint is violated, an error may be issued if it can be detected, but otherwise the result of executing the program that contains it is undefined - regardless of whether or not the inline assembly block that contains is eventually evaluated by that execution.
+
+r[asm.validity.prefix-restriction]
+Inline assembly must not end with an instruction prefix (such as `LOCK`) that would apply to instructions generated by the compiler. If this constraint is violated, an error may be issued if it can be detected, but otherwise the result of executing the program that contains it is undefined - regardless of whether or not the inline assembly block that contains is eventually evaluated by that execution.
```

Comment:
I would generalize this even more: at the end of the asm block, the CPU must be in a state where it is ready to start a new instruction. Alternatively, the asm block must end at an instruction boundary. This covers cases where you use `.byte` to only emit a part of an instruction, or leaving the assembler at an unaligned address for architectures that require aligned instructions (e.g. aarch64).